tinync/tinync.c

```diff
@@ -14,40 +14,52 @@ enum {
 struct cr {
     void *label;
     int status;
-    void *local; /* private local storage */
+    void *arg;
 };
 
-#define cr_init()                           \
-    {                                       \
-        .label = NULL, .status = CR_BLOCKED \
+#define cr_context_name(name) __cr_context_##name
+#define cr_context(name) struct cr cr_context_name(name)
+#define cr_context_init(argptr)                            \
+    {                                                      \
+        .label = NULL, .status = CR_BLOCKED, .arg = argptr \
     }
-#define cr_begin(o)                     \
-    do {                                \
-        if ((o)->status == CR_FINISHED) \
-            return;                     \
-        if ((o)->label)                 \
-            goto *(o)->label;           \
+
+#define cr_func_name(name) __cr_func_##name
+#define cr_func_def(name) static void cr_func_name(name)(struct cr * __ct)
```

Comment:
Why did you call it `__ct`?

Reply:
`__ct` stands for "context".

Reply:
Use `ctx` instead.

Reply:
The macro `cr_func_def` might be confusing. Indeed, it refers to the generator for function prototype.

Reply:
> Use `ctx` instead.

Ok, I'll update it.

> The macro `cr_func_def` might be confusing. Indeed, it refers to the generator for function prototype.

How about modiy its name as `cr_function`? It will make definition of a coroutine function become `cr_function(coroutine_name)`, which indicates itself as a coroutine function clearly.

Reply:
I intend to avoid the introduction of `cr_function` because
1. It is harder to perform static analysis such as `cppcheck` and `clang-format`.
2. Programmers might expect the unified C declarations instead of the wrapper.

Reply:
I see the point. Indeed, word `function` makes programmers think of unified C declarations.
To avoid such confusing, I think `cr_define` is a better name, which came from project [Coroutine in C](https://in4lio.github.io/ev3dev-c/group__coro.html#ga0fc518855e8bfbbd08507c5168b5179f). Renaming macro `cr_func_def` as `cr_define` will indicate the purpose of itself, which makes programmers focus on dealing with the body of coroutine, rather than declare a C function.

Reply:
Good point. You have learnt the essence of refactoring.